### PR TITLE
Distance display conditions

### PR DESCRIPTION
**Pull request by bagnell**
Adds distance display conditions to primitives.

For billboards, points and models, there is a new `distanceDisplayCondition` property. For `Primitive`s, there is a new `distanceDisplayCondition` per-instance geometry attribute.

TODO:
- [x] Add support to `PolylineCollection`.
- [x] Add support for line and path entities.
- [x] Add entity support.
- [x] Add tests for primitives.
- [x] Add tests for entities.
- [x] Add doc.
- [x] Update `CHANGES.md`.

### CODE REVIEW

**Comment by mramato** [comment]
I haven't really looked at this yet, but to go along with `translucencyByDistance` and `scaleByDistance` should this be called `showByDistance`?

**Comment by pjcozzi** [comment]
JSHint warnings: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/158843026

> Specs/Core/DistanceDisplayConditionSpec.js
  line 2  col 1  'defineSuite' is not defined.
  line 1  col 1  'definedSuite' is defined but never used.
  ⚠  2 warnings

File: Source/Scene/Model.js [hunk 1]
```diff
@@ -3719,7 +3721,8 @@ define([
             }
         }
 
-        var show = this.show && (this.scale !== 0.0);
+        var displayConditionPassed = defined(this.distanceDisplayCondition) ? this.distanceDisplayCondition.isVisible(this, frameState) : true;
+        var show = this.show && (this.scale !== 0.0) && displayConditionPassed;
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Perhaps slightly faster in the reject case (and the same speed in the accept) to do
```javascript
this.show  && displayConditionPassed && (this.scale !== 0.0)
```

File: Source/Scene/Primitive.js [hunk 2]
```diff
@@ -683,6 +683,30 @@ define([
...
+        if (!defined(primitive._attributeLocations.distanceDisplayCondition)) {
+            return vertexShaderSource;
+        }
+
+        var renamedVS = ShaderSource.replaceMain(vertexShaderSource, 'czm_non_distanceDisplayCondition_main');
+        var distanceDisplayConditionMain =
+            'attribute vec3 boundingSphereCenter3DHigh; \n' +
+            'attribute vec3 boundingSphereCenter3DLow; \n' +
+            'attribute float boundingSphereRadius; \n' +
+            'attribute vec2 distanceDisplayCondition; \n' +
```
Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Combine with `boundingSphereCenter3DHigh` and `boundingSphereCenter3DLow` and the `w` component?

**Comment by pjcozzi** [comment]
> I haven't really looked at this yet, but to go along with translucencyByDistance and scaleByDistance should this be called showByDistance?

We could if we were sure we'll never add another display condition like pixel size, viewer altitude, etc. since those would want to be named `showByPixelSize`, etc.  In practice, distance should get us 95% of the cases so up to @bagnell.

**Comment by pjcozzi** [comment]
Just those comments.  This is looking good despite all the vertex data required.

File: Source/Scene/Primitive.js [hunk 3]
```diff
@@ -683,6 +683,30 @@ define([
...
+            'void main() \n' +
+            '{ \n' +
+            '    czm_non_distanceDisplayCondition_main(); \n' +
+            '    vec4 centerRTE = czm_computeBoundingSphereCenter(); \n' +
```
Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Where is this function?  Did you forget to submit a file?

**Comment by pjcozzi** [comment]
Part of #758.  Perhaps even close that and let the KML region support stay as part of the KML roadmap.

File: Source/Scene/Primitive.js (hunk 3 above)
Comment (bagnell) [inline Source/Scene/Primitive.js]: Nope, its generated just like `czm_computePosition` for all other geometries and the functions for getting adjacent points for polylines.

Comment (pjcozzi) [inline Source/Scene/Primitive.js]: I don't see it in a repo search or this diff.

Comment (bagnell) [inline Source/Scene/Primitive.js]: See this function: https://github.com/AnalyticalGraphicsInc/cesium/blob/9efeb5a4d1ef96fbdbb9c42477db6d1da79e131b/Source/Scene/Primitive.js#L594

Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Ah, yes, thanks!

File: Source/Scene/Primitive.js (hunk 2 above)
Comment (bagnell) [inline Source/Scene/Primitive.js]: I though about that, but it doesn't fit in to how per instance geometry attributes work. Each attribute must be in its own vertex buffer so no interleaving or packing. It would be a significant amount of work to change that. We really need something similar to a 3D tiles batch table like we discussed offline.

Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Roger that.

**Comment by bagnell** [comment]
This requires the batch table from #4328, so please review and merge that PR before this one.

**Comment by pjcozzi** [comment]
@bagnell does this need another review from me?  Or just @mramato?

**Comment by bagnell** [comment]
@pjcozzi Could you review it again? Nothing has changed for billboards, points or labels, but primitives were changed to use the batch table from #4361.

File: Source/Scene/Label.js [hunk 4]
```diff
@@ -652,6 +659,30 @@ define([
...
+                //>>includeStart('debug', pragmas.debug);
+                if (defined(value) && value.far <= value.near) {
+                    throw new DeveloperError('far must be greater than near');
+                }
+                //>>includeEnd('debug');
+                if (!DistanceDisplayCondition.equals(value, this._distanceDisplayCondition)) {
+                    this._distanceDisplayCondition = DistanceDisplayCondition.clone(value, this._distanceDisplayCondition);
+
+                    var glyphs = this._glyphs;
+                    for (var i = 0, len = glyphs.length; i < len; i++) {
```
Comment (pjcozzi) [inline Source/Scene/Label.js]: Have you ran into any scenarios, perhaps with a large font size, where only some letters in a string pass/fail the distance test?

File: Source/Scene/Model.js [hunk 5]
```diff
@@ -859,6 +863,22 @@ define([
...
+            get : function() {
+                return this._distanceDisplayCondition;
+            },
+            set : function(value) {
+                //>>includeStart('debug', pragmas.debug);
+                if (defined(value) && value.far <= value.near) {
+                    throw new DeveloperError('far must be greater than near');
+                }
+                //>>includeEnd('debug');
+                if (!DistanceDisplayCondition.equals(value, this._distanceDisplayCondition)) {
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: This check doesn't really add value here, does it?

File: Source/Scene/Polyline.js [hunk 6]
```diff
@@ -285,6 +289,23 @@ define([
...
+        },
+
+        distanceDisplayCondition : {
+            get : function() {
+                return this._distanceDisplayCondition;
+            },
+            set : function(value) {
+                if (!DistanceDisplayCondition.equals(value, this._distanceDisplayCondition)) {
+                    //>>includeStart('debug', pragmas.debug);
+                    if (defined(value) && value.far <= value.near) {
```
Comment (pjcozzi) [inline Source/Scene/Polyline.js]: To avoid side effects, this check should be outside if the above `if` statement, right?

**Comment by pjcozzi** [comment]
Optional, but it would be nice if the Sandcastle examples automatically zoomed from a home view to the object so that the distance condition is obvious, especially point/model, which requires a lot of zooming.

**Comment by pjcozzi** [comment]
Just those comments, would like to also read the reference doc when ready.

File: Source/Scene/Label.js (hunk 4 above)
Comment (bagnell) [inline Source/Scene/Label.js]: No. All of the glyphs in the label use the same position relative to the eye when comparing the distance. Do you have an example where it happens?

**Comment by bagnell** [comment]
@pjcozzi The reference doc is ready.

**Comment by pjcozzi** [comment]
Are you sure that we don't want `distanceDisplayCondition` to be part of the `Model` constructor (and billboard/label add just like `scaleByDistance`)?

**Comment by pjcozzi** [comment]
Just those comments.

**Comment by bagnell** [comment]
>Are you sure that we don't want distanceDisplayCondition to be part of the Model constructor (and billboard/label add just like scaleByDistance)?

It is part of the constructors. The doc was missing for `Model`, but there is no doc for billboard/label options either on the constructor or on add.

**Comment by pjcozzi** [comment]
All good with me.  @mramato please merge when ready.  The changes to the entity layer are minor.

File: Source/Core/DistanceDisplayCondition.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+         * @type {Number}
+         * @default 0.0
+         */
+        near : {
+            get : function() {
+                return this._near;
+            },
+            set : function(value) {
+                this._near = value;
+                this._near2 = value * value;
```
Comment (mramato) [inline Source/Core/DistanceDisplayCondition.js]: Nitpicky, but as a generate practice, we would normally call this `nearSquared`.

File: Source/Core/DistanceDisplayCondition.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    };
+
+    /**
+     * Duplicates a distance display condition instance.
+     *
+     * @param {DistanceDisplayCondition} [value] The distance display condition to duplicate.
+     * @param {DistanceDisplayCondition} [result] The result onto which to store the result.
+     * @return {DistanceDisplayCondition} The duplicated instance.
+     */
+    DistanceDisplayCondition.clone = function(value, result) {
```
Comment (mramato) [inline Source/Core/DistanceDisplayCondition.js]: Since this is used directly by the Entity api as well, it should have a prototype clone too.  Right now the Entity API ends up making a new instance every time get value is called because no prototype clone is defined.

**Comment by mramato** [comment]
Point & Model does not seem to work at all in 2D or Columbus View.

**Comment by mramato** [comment]
Actually, I wasn't zoomed in enough.  It appears to partially work but the model is never drawn.

**Comment by mramato** [comment]
This may be in master too, so create an issue if it's not related to this branch.
1. Run the Sandcastle example.
2. Switch to 2D
3. Double click the point
4. Switch to 3D. (The morph is all screwy)

File: Source/DataSources/EllipsoidGeometryUpdater.js [hunk 9]
```diff
@@ -649,7 +679,8 @@ define([
                     modelMatrix : !in3D ? modelMatrix : undefined,
                     attributes : {
                         show : new ShowGeometryInstanceAttribute(showOutline),
-                        color : ColorGeometryInstanceAttribute.fromColor(outlineColor)
+                        color : ColorGeometryInstanceAttribute.fromColor(outlineColor),
+                        distanceDisplayCondition : distanceDisplayConditionAttribute
```
Comment (mramato) [inline Source/DataSources/EllipsoidGeometryUpdater.js]: This needs to be handled in the else path as well, just like show and color.

**Comment by mramato** [comment]
It doesn't look like any of the `DynamicGeometryUpdater` objects have been updated to set the displayCondition property on the underlying primitives.  These are private classes that are part of each GeometryUpdater, just search for `DynamicGeometryUpdater` and you'll find them.

For the above, a corresponding test needs to be added to the `dynamic updater sets properties` test of each updated (again, searching for that phrase should reveal them all).

Basically, it looks like dynamic geometry never gets its display condition updated.

**Comment by mramato** [comment]
That's all I have for now.  Thanks @bagnell

**Comment by bagnell** [comment]
>This may be in master too, so create an issue if it's not related to this branch.
1. Run the Sandcastle example.
2. Switch to 2D
3. Double click the point
4. Switch to 3D. (The morph is all screwy)

Models in 2D crash in master so I'll open an issue for it. In this branch the same weird morph happens for the model example.

**Comment by bagnell** [comment]
>Basically, it looks like dynamic geometry never gets its display condition updated.

All the dynamic geometries do update the distance display condition. It was only missing from that else block you mentioned above.

>For the above, a corresponding test needs to be added to the dynamic updater sets properties test of each updated (again, searching for that phrase should reveal them all).

I didn't modify the test because it doesn't change any per-instance attributes besides show. Then, it makes sure that a primitive is created when true.

@mramato This is ready for another look. I'll add an issue for the model crash in 2D.

**Comment by mramato** [comment]
`Scene/GroundPrimitive renders with distance display condition per instance attribute` is failing on my machine.

**Comment by bagnell** [comment]
@mramato The test is fixed. It wasn't updated with the other tests that were failing in master.

**Comment by mramato** [comment]
👍
